src/services/user/hooks/registrationPins.js

```diff
@@ -14,43 +15,63 @@ const generatePin = (hook) => {
 };
 
 function createinfoText(hook) {
-	let text;
 	const role = hook.data.mailTextForRole;
 	const { pin } = hook.data;
 	const shortTitle = process.env.SC_SHORT_TITLE || 'Schul-Cloud*';
 	const longTitle = process.env.SC_TITLE || 'HPI Schul-Cloud*';
-	if (!role || !pin) {
-		logger.warning('Role or PIN missing to define mail text.');
-		return '';
+	if (!pin) {
+		throw new BadRequest('Fehler beim Erstellen der Pin.');
 	}
 	if (role === 'parent') {
-		text = `Vielen Dank, dass Sie Ihrem Kind durch Ihr Einverständnis die Nutzung der ${longTitle} ermöglichen.
-Bitte geben Sie folgenden Code ein, wenn Sie danach gefragt werden, um die Registrierung abzuschließen:
+		return `Vielen Dank, dass Sie Ihrem Kind durch Ihr Einverständnis die Nutzung der ${longTitle} ermöglichen.
+Bitte geben Sie den folgenden Bestätigungscode im Registrierungsprozess ein, um die Registrierung abzuschließen:
 
 PIN: ${pin}
 
 Mit Freundlichen Grüßen
 Ihr ${shortTitle} Team`;
-	} else if (role === 'student' || role === 'employee' || role === 'expert') {
-		text = `Vielen Dank, dass du die ${longTitle} nutzen möchtest.
-Bitte gib folgenden Code ein, wenn du danach gefragt wirst, um die Registrierung abzuschließen:
+	}
+	if (role === 'student' || role === 'employee' || role === 'expert') {
+		return `Vielen Dank, dass du die ${longTitle} nutzen möchtest.
+Bitte gib den folgenden Bestätigungscode im Registrierungsprozess ein, um deine Registrierung bei der ${longTitle} abzuschließen:
 
 PIN: ${pin}
 
 Mit freundlichen Grüßen
 Dein ${shortTitle} Team`;
-	} else {
-		logger.warning('No valid role submitted to define mail text.');
-		return '';
 	}
-	return text;
+	throw new BadRequest('Die angegebene Rolle ist ungültig.', { role });
 }
 
 const checkAndVerifyPin = (hook) => {
-	if (hook.result.data.length === 1 && hook.result.data[0].verified === false) {
-		return hook.app.service('registrationPins').patch(hook.result.data[0]._id, { verified: true }).then(() => Promise.resolve(hook));
+	if (hook.result.data.length === 0) {
+		return hook;
 	}
-	return Promise.resolve(hook);
+	if (hook.result.data.length === 1) {
+		const firstDataItem = hook.result.data[0];
+		// check generation age
+		const now = Date.now();
+		if (firstDataItem.updatedAt.getTime() + (Configuration.get('PIN_MAX_AGE_SECONDS') * 1000) < now) {
+			throw new Forbidden('Der eingegebene Code ist nicht mehr gültig. Bitte fordere einen neuen Code an.');
+		}
+		if (firstDataItem.verified === true) {
+			// already verified
+			return hook;
+		}
+		if (firstDataItem.pin) {
+			if (firstDataItem.pin === hook.params.query.pin) {
+				return hook.app.service('registrationPins')
+					.patch(firstDataItem._id, { verified: true })
+					.then((result) => {
+						hook.result.data = [result];
+						return hook;
+					});
+			}
+			throw new BadRequest('Der eingegebene Code ist ungültig oder konnte nicht bestätigt werden. Bitte versuche es erneut.');
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [This line has a length of 132. Maximum allowed is 120.](https://app.codacy.com/manual/schul-cloud/schulcloud-server/pullRequest?prid=5203912)